test/es-module/test-esm-resolve-type.js

```diff
@@ -0,0 +1,192 @@
+      fs.rmdirSync(nmDir);
+    });
+  })().then(common.mustCall());
+
+  // Helpers
+  const createDir = (path) => {
+    if (!fs.existsSync(path)) {
+      fs.mkdirSync(path);
+    }
+  };
+
+  // Create a dummy dual package
+  //
+  /**
+   * this creates following directory structure:
+   *
+   * ./node_modules:
+   *   |-> my-dual-package
+   *       |-> es
+   *           |-> index.js
+   *           |-> package.json [2]
+   *       |-> lib
+   *           |-> index.js
+   *       |->package.json [1]
+   *
+   * [1] - main package.json of the package
+   *     - it contains:
+   *             - type: 'commonjs'
+   *             - main: 'lib/mainfile.js'
+   *             - conditional exports for 'require' (lib/index.js) and
+   *                                       'import' (es/index.js)
+   * [2] - package.json add-on for the import case
+   *     - it only contains:
+   *             - type: 'module'
+   *
+   * in case the package is consumed as an ESM by importing it:
+   *    import * as my-package from 'my-dual-package'
+   * it will cause the resolve method to return:
+   *  {
+   *     url: '<base_path>/node_modules/my-dual-package/es/index.js',
+   *     format: 'module'
+   *  }
+   *
+   *  following testcase ensures that resolve works correctly in this case
+   *  returning the information as specified above. Source for 'url' value
+   *  is [1], source for 'format' value is [2]
+   */
+
+  const moduleName = 'my-dual-package';
+
+  const nmDir = rel('node_modules');
+  const mDir = rel(`node_modules/${moduleName}`);
+  const esSubDir = rel(`node_modules/${moduleName}/es`);
+  const cjsSubDir = rel(`node_modules/${moduleName}/lib`);
+  const pkg = rel(`node_modules/${moduleName}/package.json`);
+  const esmPkg = rel(`node_modules/${moduleName}/es/package.json`);
+  const esScript = rel(`node_modules/${moduleName}/es/index.js`);
+  const cjsScript = rel(`node_modules/${moduleName}/lib/index.js`);
+
+  createDir(nmDir);
+  createDir(mDir);
+  createDir(esSubDir);
+  createDir(cjsSubDir);
+
+  const mainPkgJsonContent = {
+    type: 'commonjs',
+    main: 'lib/index.js',
+    exports: {
+      '.': {
+        'require': './lib/index.js',
+        'import': './es/index.js'
+      },
+      './package.json': './package.json',
+    }
+  };
+  const esmPkgJsonContent = {
+    type: 'module'
+  };
+
+  fs.writeFileSync(pkg, JSON.stringify(mainPkgJsonContent));
+  fs.writeFileSync(esmPkg, JSON.stringify(esmPkgJsonContent));
+  fs.writeFileSync(esScript,
+                   'export function esm-resolve-tester() {return 42}');
+  fs.writeFileSync(cjsScript,
+                   `module.exports = { 
+                     esm-resolve-tester: () => {return 42}}`
+  );
+
+
+  // test the resolve
+  const resolveResult = resolve(`${moduleName}`);
+  assert.strictEqual(resolveResult.format, 'module');
+  assert.ok(resolveResult.url.includes('my-dual-package/es/index.js'));
+
+  // cleanup
+  fs.unlinkSync(esScript);
```

Comment:
Could we move the cleanup also into the finally block to avoid leftovers in case of a failed test.
Maybe deleting the whole `tmpdir` recursive would be easier?

Reply:
that is a good point, makes it cleaner. I pushed the change in https://github.com/nodejs/node/pull/40980/commits/33a457247c1d7ddb8b59d9ce435b99c2353cf270